File: modules/protocol/base/src/main/java/candle/protocol/codec/TypeCodec.java
package candle.protocol.codec;

import candle.types.resource.NamespacedObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public abstract class TypeCodec<T> implements NamespacedObject {
  abstract T read( InputStream in, AbstractDataCodec dataCodec ) throws
                                                                 IOException;

  abstract void write( OutputStream out, T value, AbstractDataCodec dataCodec ) throws
                                                                                IOException;

  public TypeIdentifier<T> identifier( String namespacedIdentifier ) {
    String[] splitIdentifier = namespacedIdentifier.split(":", 1);
    return new TypeIdentifier<>(splitIdentifier[0], splitIdentifier[1], this);
  }

  public TypeIdentifier<T> identifier( String namespace, String identifier ) {
    return new TypeIdentifier<>(namespace, identifier, this);
  }
}

File: protocol/base/src/main/kotlin/net/candlemc/protocol/codec/TypeCodec.kt
package net.candlemc.protocol.codec

import net.candlemc.types.resource.NamespacedObject
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

abstract class TypeCodec<T> : NamespacedObject {
    @Throws(IOException::class)
    abstract fun read(inputStream: InputStream, dataCodec: AbstractDataCodec): T

    @Throws(IOException::class)
    abstract fun write(outputStream: OutputStream, value: T, dataCodec: AbstractDataCodec)

    fun identifier(namespacedIdentifier: String): TypeIdentifier<T> {
        val splitIdentifier = namespacedIdentifier.split(":".toRegex(), limit = 1).toTypedArray()
        return TypeIdentifier<T>(splitIdentifier[0], splitIdentifier[1], this)
    }

    fun identifier(namespace: String, identifier: String): TypeIdentifier<T> {
        return TypeIdentifier<T>(namespace, identifier, this)
    }
}
